projects/vdk-core/src/vdk/internal/builtin_plugins/connection/execution_cursor.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+from typing import Any
+
+from vdk.internal.builtin_plugins.connection.decoration_cursor import ManagedOperation
+from vdk.internal.builtin_plugins.connection.pep249.interfaces import PEP249Cursor
+
+
+class ExecuteOperationResult:
+    def __init__(self, native_result: Any):
+        self.__native_result = native_result
+
+    def get_native_result(self) -> Any:
+        """
+        Returns the result of PEP249Cursor.execute invocation using the native cursor.
+        Since this is vendor specific the result type is any - but usually it's number of selected or updated rows.
+        :return: Any
+        """
+        return self.__native_result
+
+
+class ExecutionCursor(PEP249Cursor):
+    """
+    Extends PEP249Cursor to provide:
+        * ability to directly access and execute operations with the native cursor.
+        Generally it should be used only if default implementation does not work (which should be almost never)
+         or more likely - to use some vendor specific features.
+
+    See connection_hook_spec#db_connection_execute_operation for more details and examples how to use it.
+    """
+
+    def __init__(
+        self,
+        native_cursor: PEP249Cursor,
+        managed_operation: ManagedOperation,
+        log=logging.getLogger(__name__),
+    ):
+        super().__init__(native_cursor, log)
+        self.__managed_operation = managed_operation
+        self.__native_cursor = native_cursor
+
+    def get_native_cursor(self) -> PEP249Cursor:
+        """
+        Get the underlying native cursor. Used to actually execute the operation.
+        Generally should not be accessed directly.
+        But it's useful to be exposed in case the native library used provide some extra features.
+        For example Impala (impyla driver) provides a way to get the profile of a query after it's executed
+        which can be printed to aid debugging.
+
+        Check the example usages in docstring of connection_hook_spec#db_connection_execute_operation .
+
+        :return: PEP249Cursor
+            the underlying native cursor being managed.
```

Comment:
what is the use-case for publicly exposing the native cursor? execution cursor should be itself able to handle everything the native does, and more. Why running away from being on the data path design?

Reply:
I think I explained the use-cases before: 
 

There are compelling use-cases for access to native/raw cursor by users as it provides extra features. 

Impala (impyla driver) provides a way to get the profile of a query after it's executed - this is something that vdk-impala plugin can use to extract lineage and statistics
Trino driver provides similar way to get information about the query from the raw cursor - again for similar reasons - to extract lineage and stats.

Those are implementation-specific but necessary to achieve feature-rich database plugins.

Those are important use-cases, for example, I want to get extra debug info this with Impala driver:
```
@hookimpl(hookwrapper=True)
            db_connection_execute_operation(execution_cursor: ExecutionCursor) -> Optional[int]:
                yield # let the query execute first
                c = cast(impala.interface.Cursor, execution_cursor.get_native_cursor())
                log.info(f"Query {execution_cursor.get_managed_operation().get_operation()} debug info:"
                         f"summary: {c.get_summary()}, profile: {c.get_profile()}")
```

For now this is the best way I have figured out how to do that. 
I have a another idea - If we reuse the functionality in [ManagedCursor](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/connection/managed_cursor.py#L39) for  Dynamic interception and delegation , then we may not need get_native_cursor method. 

I will do that in separate PR (after merging this one) though.